File: _includes/code/java-v6/src/test/java/SearchAggregateTest.java
import io.weaviate.client6.v1.api.WeaviateClient;
import io.weaviate.client6.v1.api.collections.CollectionHandle;
import io.weaviate.client6.v1.api.collections.aggregate.Aggregate;
import io.weaviate.client6.v1.api.collections.aggregate.GroupBy;
import io.weaviate.client6.v1.api.collections.query.Where;
import org.junit.jupiter.api.AfterAll;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Test;

import java.io.IOException;
import java.util.Map;

class SearchAggregateTest {

  private static WeaviateClient client;

  @BeforeAll
  public static void beforeAll() throws IOException {
    // START INSTANTIATION-COMMON
    // Best practice: store your credentials in environment variables
    String weaviateUrl = System.getenv("WEAVIATE_URL");
    String weaviateApiKey = System.getenv("WEAVIATE_API_KEY");
    String openaiApiKey = System.getenv("OPENAI_APIKEY");

    client = WeaviateClient.connectToWeaviateCloud(
        weaviateUrl,
        weaviateApiKey,
        config -> config.setHeaders(Map.of("X-OpenAI-Api-Key", openaiApiKey)));
    // END INSTANTIATION-COMMON
  }

  @AfterAll
  public static void afterAll() throws Exception {
    client.close();
  }

  @Test
  void testMetaCount() {
    // START MetaCount Python
    CollectionHandle<Map<String, Object>> jeopardy = client.collections.use("JeopardyQuestion");
    var response = jeopardy.aggregate.overAll(
        // highlight-start
        a -> a.includeTotalCount(true)
    // highlight-end
    );

    System.out.println(response.totalCount());
    // END MetaCount Python
  }

  @Test
  void testTextProp() {
    // START TextProp Python
    CollectionHandle<Map<String, Object>> jeopardy = client.collections.use("JeopardyQuestion");
    var response = jeopardy.aggregate.overAll(
        // TODO[g-despot] Count, value and min occurences?
        // highlight-start
        a -> a.metrics(
            Aggregate.text("answer", m -> m
                .topOccurences()
                .topOccurencesCutoff(5) // Threshold minimum count
            ))
    // highlight-end
    );
    // TODOπ[g-despot] How to get topOccurences here
    System.out.println(response.properties().get("answer"));
    // END TextProp Python
  }

  @Test
  void testIntProp() {
    // START IntProp Python
    CollectionHandle<Map<String, Object>> jeopardy = client.collections.use("JeopardyQuestion");
    var response = jeopardy.aggregate.overAll(
        // highlight-start
        // Use .number for floats (NUMBER datatype in Weaviate)
        a -> a.metrics(
            Aggregate.integer("points", m -> m
                .sum()
                .max()
                .min()))
    // highlight-end
    );

    // TODOπ[g-despot] How to get sum, min and max here
    System.out.println(response.properties().get("points"));
    System.out.println(response.properties().get("points"));
    System.out.println(response.properties().get("points"));
    // END IntProp Python
  }

  @Test
  void testGroupBy() {
    // START groupBy Python
    CollectionHandle<Map<String, Object>> jeopardy = client.collections.use("JeopardyQuestion");
    var response = jeopardy.aggregate.overAll(
        // TODO[g-despot] Why is metrics needed here?
        // highlight-start
        c -> c.metrics(),
        GroupBy.property("round")
    // highlight-end
    );

    // print rounds names and the count for each
    for (var group : response.groups()) {
      System.out.printf("Value: %s Count: %d\n", group.groupedBy().value(), group.totalCount());
    }
    // END groupBy Python
  }

  @Test
  void testNearTextWithLimit() {
    // START nearTextWithLimit Python
    CollectionHandle<Map<String, Object>> jeopardy = client.collections.use("JeopardyQuestion");
    var response = jeopardy.aggregate.nearText(
        "animals in space",
        a -> a
            // highlight-start
            .objectLimit(10)
            // highlight-end
            .metrics(Aggregate.number("points", m -> m.sum())));

    System.out.println(response.properties().get("points"));
    // END nearTextWithLimit Python
  }

  @Test
  void testHybrid() {
    // START HybridExample
    CollectionHandle<Map<String, Object>> jeopardy = client.collections.use("JeopardyQuestion");
    var response = jeopardy.aggregate.hybrid(
        "animals in space",
        a -> a
            // TODO what about bm25Operator?
            // .bm25Operator(...) // Additional parameters available, such as
            // `bm25_operator`, `filter` etc.
            // highlight-start
            .objectLimit(10)
            // highlight-end
            .metrics(Aggregate.number("points", m -> m.sum())));

    System.out.println(response.properties().get("points"));
    // END HybridExample
  }

  @Test
  void testNearTextWithDistance() {
    // START nearTextWithDistance Python
    CollectionHandle<Map<String, Object>> jeopardy = client.collections.use("JeopardyQuestion");
    var response = jeopardy.aggregate.nearText(
      // TODO[g-despot] Should be distance instead of objectLimit
        "animals in space",
        a -> a.objectLimit(10)
            // highlight-start
            //.distance(0.19f)
            // highlight-end
            .metrics(Aggregate.number("points", m -> m.sum())));

    System.out.println(response.properties().get("points"));
    // END nearTextWithDistance Python
  }

  @Test
  void testWhereFilter() {
    // START whereFilter Python
    // TODO[g-despot] Why is where not available on overAll()?
    CollectionHandle<Map<String, Object>> jeopardy = client.collections.use("JeopardyQuestion");
    var response = jeopardy.aggregate.overAll(
        a -> a
            // highlight-start
            // .where(Where.property("round").eq("Final Jeopardy!"))
            // highlight-end
            .includeTotalCount(true));

    System.out.println(response.totalCount());
    // END whereFilter Python
  }
}

File: _includes/code/csharp/SearchAggregateTest.cs
using System;
using System.Collections.Generic;
using System.Text.Json;
using System.Threading.Tasks;
using Weaviate.Client;
using Weaviate.Client.Models;
using Xunit;

namespace WeaviateProject.Tests;

public class SearchAggregateTest : IDisposable
{
    private static readonly WeaviateClient client;

    // Static constructor for one-time setup (like @BeforeAll)
    static SearchAggregateTest()
    {
        // START INSTANTIATION-COMMON
        // Best practice: store your credentials in environment variables
        string weaviateUrl = Environment.GetEnvironmentVariable("WEAVIATE_URL");
        string weaviateApiKey = Environment.GetEnvironmentVariable("WEAVIATE_API_KEY");
        string openaiApiKey = Environment.GetEnvironmentVariable("OPENAI_API_KEY");

        client = Connect
            .Cloud(
                weaviateUrl,
                weaviateApiKey,
                headers: new Dictionary<string, string>() { { "X-OpenAI-Api-Key", openaiApiKey } }
            )
            .GetAwaiter()
            .GetResult();
        // END INSTANTIATION-COMMON
    }

    // Dispose is called once after all tests in the class are finished (like @AfterAll)
    public void Dispose()
    {
        // The C# client manages connections automatically and does not require an explicit 'close' method.
        GC.SuppressFinalize(this);
    }

    [Fact]
    public async Task TestMetaCount()
    {
        // START MetaCount
        var jeopardy = client.Collections.Use("JeopardyQuestion");
        var response = await jeopardy.Aggregate.OverAll(
            // highlight-start
            totalCount: true
        // highlight-end
        );

        Console.WriteLine(response.TotalCount);
        // END MetaCount
    }

    [Fact]
    public async Task TestTextProp()
    {
        // START TextProp
        var jeopardy = client.Collections.Use("JeopardyQuestion");
        var response = await jeopardy.Aggregate.OverAll(
            // highlight-start
            returnMetrics:
            [
                Metrics
                    .ForProperty("answer")
                    .Text(
                        topOccurrencesCount: true,
                        topOccurrencesValue: true,
                        minOccurrences: 5 // Threshold minimum count
                    ),
            ]
        // highlight-end
        );

        var answerMetrics = response.Properties["answer"] as Aggregate.Text;
        if (answerMetrics != null)
        {
            Console.WriteLine(JsonSerializer.Serialize(answerMetrics.TopOccurrences));
        }
        // END TextProp
    }

    [Fact]
    public async Task TestIntProp()
    {
        // START IntProp
        var jeopardy = client.Collections.Use("JeopardyQuestion");
        var response = await jeopardy.Aggregate.OverAll(
            // highlight-start
            // Use .Number for floats (NUMBER datatype in Weaviate)
            returnMetrics:
            [
                Metrics.ForProperty("points").Integer(sum: true, maximum: true, minimum: true),
            ]
        // highlight-end
        );

        var pointsMetrics = response.Properties["points"] as Aggregate.Integer;
        if (pointsMetrics != null)
        {
            Console.WriteLine($"Sum: {pointsMetrics.Sum}");
            Console.WriteLine($"Max: {pointsMetrics.Maximum}");
            Console.WriteLine($"Min: {pointsMetrics.Minimum}");
        }
        // END IntProp
    }

    [Fact]
    public async Task TestGroupBy()
    {
        // START groupBy
        var jeopardy = client.Collections.Use("JeopardyQuestion");
        var response = await jeopardy.Aggregate.OverAll(
            // highlight-start
            groupBy: new Aggregate.GroupBy("round")
        // highlight-end
        );

        // print rounds names and the count for each
        foreach (var group in response.Groups)
        {
            Console.WriteLine($"Value: {group.GroupedBy.Value} Count: {group.TotalCount}");
        }
        // END groupBy
    }

    [Fact]
    public async Task TestNearTextWithLimit()
    {
        // START nearTextWithLimit
        var jeopardy = client.Collections.Use("JeopardyQuestion");
        var response = await jeopardy.Aggregate.NearText(
            "animals in space",
            // highlight-start
            limit: 10,
            // highlight-end
            returnMetrics: [Metrics.ForProperty("points").Number(sum: true)]
        );

        var pointsMetrics = response.Properties["points"] as Aggregate.Number;
        Console.WriteLine(JsonSerializer.Serialize(pointsMetrics));
        // END nearTextWithLimit
    }

    [Fact]
    public async Task TestHybrid()
    {
        // START HybridExample
        var jeopardy = client.Collections.Use("JeopardyQuestion");
        var response = await jeopardy.Aggregate.Hybrid(
            "animals in space",
            // Additional parameters are available, such as `bm25Operator`, `filters`, etc.
            // highlight-start
            objectLimit: 10,
            // highlight-end
            returnMetrics: [Metrics.ForProperty("points").Number(sum: true)]
        );

        var pointsMetrics = response.Properties["points"] as Aggregate.Number;
        Console.WriteLine(JsonSerializer.Serialize(pointsMetrics));
        // END HybridExample
    }

    [Fact]
    public async Task TestNearTextWithDistance()
    {
        // Note: The C# client supports the 'distance' parameter for nearText aggregations.
        // START nearTextWithDistance
        var jeopardy = client.Collections.Use("JeopardyQuestion");
        var response = await jeopardy.Aggregate.NearText(
            ["animals in space"],
            // highlight-start
            distance: 0.19,
            // highlight-end
            returnMetrics: [Metrics.ForProperty("points").Number(sum: true)]
        );

        var pointsMetrics = response.Properties["points"] as Aggregate.Number;
        Console.WriteLine(JsonSerializer.Serialize(pointsMetrics));
        // END nearTextWithDistance
    }

    [Fact]
    public async Task TestWhereFilter()
    {
        // Note: The C# client supports the 'filter' parameter for OverAll aggregations.
        // START whereFilter
        var jeopardy = client.Collections.Use("JeopardyQuestion");
        var response = await jeopardy.Aggregate.OverAll(
            // highlight-start
            filters: Filter.Property("round").IsEqual("Final Jeopardy!"),
            // highlight-end
            totalCount: true
        );

        Console.WriteLine(response.TotalCount);
        // END whereFilter
    }
}
